rust/kernel/cpu.rs

```diff
@@ -0,0 +1,131 @@
+//!
+//! Note: This module is a work in progress.
+
+use crate::{bindings, error::code, prelude::*};
+use core::{marker::PhantomData, mem::ManuallyDrop};
+
+/// Represents an SMP CPU by its zero-based kernel processor id.
+///
+/// Cpu values can be safely passed between threads (and across physical CPUs) as they are
+/// merely an integer index representing a CPU. All actual interaction with CPUs is late-
+/// bound.
+pub struct Cpu(i32);
+
+struct FnCall<F, R>
+where
+    F: FnOnce() -> R,
+{
+    out: Option<R>,
+    func: ManuallyDrop<F>,
+}
+
+unsafe extern "C" fn call_fn<F, R>(fn_box: *mut core::ffi::c_void)
+where
+    F: FnOnce() -> R,
+{
+    let fn_call = unsafe { &mut *(fn_box as *mut FnCall<F, R>) };
+    let func = unsafe { ManuallyDrop::take(&mut fn_call.func) };
+    fn_call.out = Some(func());
+}
+
+impl Cpu {
+    /// Constructs a new instance referring to a specific processor by id.
+    ///
+    /// # Safety
+    ///
+    /// This function checks that the passed id is a valid _potential_ processor id, but does
+    /// not check that the processor is actually present or online. Manipulation of absent
+    /// or offline processors will fail with reasonable errors returned.
+    pub fn new(id: i32) -> Result<Cpu> {
+        if id >= Self::num_possible() {
+            return Err(code::E2BIG);
+        } else if id < 0 {
+            return Err(code::EINVAL);
+        }
+        Ok(Cpu(id))
+    }
+
+    /// Returns the number of processors that are possible (including onlining and hotplug).
+    pub fn num_possible() -> i32 {
+        // SAFETY: The number of possible CPUs is fixed early at boot and does not change in
+        // response to processors coming online, offline, or hotplug events.
+        unsafe { bindings::nr_cpu_ids as i32 }
+    }
+
+    /// Lock the running thread to the current CPU and prevent preemption.
+    pub fn lock_current() -> Guard {
+        Guard::new()
+    }
+
+    /// Get the processor id of this CPU.
+    pub fn id(&self) -> i32 {
+        self.0
+    }
+
+    /// Call a function on the CPU identified by `self`.
+    pub fn call<R, F>(&self, func: F) -> R
+    where
+        F: FnOnce() -> R + Send,
+        R: Send,
+    {
+        // Evaluate func locally if we're already on the right CPU.
+        {
+            let guard = Self::lock_current();
+            if self.0 == guard.cpu().id() {
+                return func();
+            }
+        }
+
+        let mut fn_call = FnCall {
+            out: None,
+            func: ManuallyDrop::new(func),
+        };
+        // SAFETY: smp_call_function_single is invoked with wait == 1 (true) ensuring that fn_call
+        // outlives this FFI invocation. func's returned value is moved into fn_call.out and then
+        // returned here. The Send type constraint ensures func and its returned value can safely
+        // be moved between threads/CPUs.
+        unsafe {
+            bindings::smp_call_function_single(
```

Comment:
`smp_call_function_single` may fail, for example, if the CPU is not online.

We need to check the return value so that we can report this to the caller (which means that you probably want to return `Result<R>`). We also can't assume that `fn_call.out` has been initialised and `fn_call.func` has been dropped if `smp_call_function_single` fails.

Reply:
This was a silly miss in the original PR. Fixed.